### PR TITLE
[Inductor] Fix Argmax codegen with Nan input

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118358

**Summary**
Fix issue https://github.com/pytorch/pytorch/issues/118266, current `torch.argmax` and `torch.argmin` has different return values with eager and Inductor cpp backend when inputs has `Nan` value. Align cpp backend results to eager by reusing the compare function.

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_argmin_cpu_only
python -u -m pytest -s -v test_cpu_repro.py -k test_argmax_argmin_with_nan_value
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by lezcano — APPROVED** [review]

**Comment by lezcano** [comment]
The errors are real tho.

**Comment by leslie-fang-intel** [comment]
> The errors are real tho.

Yean, but I am not able to reproduce in my local dev system.

File: torch/_inductor/codegen/cpp_prefix.h [hunk 1]
```diff
@@ -92,6 +93,18 @@ Welford<T> welford_combine(const Welford<T> &acc, T data) {
   return result;
 }
 
+// Refer to https://github.com/pytorch/pytorch/blob/b5b36cf0c4e1958f1ff25120f5d4beeef3288187/
+// aten/src/ATen/native/SharedReduceOps.h#L419-L445
+template <typename scalar_t>
+inline bool greater_or_nan(scalar_t a, scalar_t b, int64_t idx_a, int64_t idx_b) {
+  // TODO<Leslie> Should we include the header file of <ATen/native/SharedReduceOps.h> or copy the implementation
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:100]: Considering the implementation is not that complex, maybe directly implement them here is better than including another header file?

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Review by jansel — APPROVED** [review]

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -1974,6 +1974,58 @@ def fn(x):
                 self.common(fn, (x,))
                 assert metrics.generated_cpp_vec_kernel_count == 0
 
+    @unittest.skipIf(
+        not codecache.valid_vec_isa_list(), "Does not support vectorization"
+    )
+    @patch("torch.cuda.is_available", lambda: False)
+    def test_argmin_cpu_only(self):
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_repro.py:1981]: Thanks for the comment.

> It is in test_cpu_repo.py, no need to mark to cpu_only and check cuda here?

Remove the `cpu_only` postfix and cuda check here.

> BTW, should the basic argmin functionality already be covered somewhere else?

I didn't find `argmin` tested in the `test_cpu_repo.py`, so add one in this test file.

File: test/inductor/test_cpu_repro.py [hunk 3]
```diff
@@ -1974,6 +1974,58 @@ def fn(x):
...
+            with config.patch({"cpp.simdlen": None}):
+                torch._dynamo.reset()
+                metrics.reset()
+                self.common(fn, (x,))
+                assert metrics.generated_cpp_vec_kernel_count == 0
+
+    @unittest.skipIf(
+        not codecache.valid_vec_isa_list(), "Does not support vectorization"
+    )
+    @patch("torch.cuda.is_available", lambda: False)
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_repro.py:1997]: Changed.

File: test/inductor/test_cpu_repro.py [hunk 4]
```diff
@@ -1974,6 +1974,58 @@ def fn(x):
...
+            torch.Tensor([-755832.1250, 100, 200]),
+            torch.Tensor([100, -755832.1250]),
+            torch.Tensor([100, 200, -755832.1250]),
+        ]
+
+        for x in inputs:
+            x = x.repeat(16, 16)
+            x = torch.log1p(x)
+
+            with config.patch({"cpp.simdlen": None}):
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_repro.py:2016]: Thanks for the comment and removed.

File: torch/_inductor/codegen/cpp_prefix.h (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_prefix.h]: Thanks for the comment and changed.

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**